### ISSUE TITLE
About the softmax layer & label process?

### ISSUE THREAD

**Issue by YanZhiyuan0918** [bonlime/keras-deeplab-v3-plus#54]
Hi, bonlime,
Thanks for your work! It helps me a lot! But I have two questions in my training process:

1. I found there is no 'softmax' on the last layer of the Deeplabv3() model. Is it right if I use model.fit() directly after model=Deeplabv3() without any other process? Or could you please tell me where to use the 'softmax' layer?

2. In the training image input process, how to correctly deal with labels? In my previous segmentation projects, the training and prediction steps usually operate on single-channel pngs labels, where the value of each pixel corresponds to the class label (for example, 0-21 for the Pascal dataset). Is this project the same as that?

I'm looking forward to your reply. Thanks!

**Comment by JeniaNovellusDx** [comment]
I also think that the softmax is missing, I've gotten strange nan values in the accuracy\loss and then I realized that the softmax is missing!

**Comment by YanZhiyuan0918** [comment]
@JeniaNovellusDx  hi, have you added the softmax layer? What about the experiment effect?

**Comment by fengsam** [comment]
you can fusion softmax to loss。
def categorical_crossentropy(y_true, y_pred):
    return K.categorical_crossentropy(y_true, y_pred, from_logits=True)

**Comment by Meight** [comment]
@YanZhiyuan0918 the softmax layer isn't present because the model from the paper also outputs logits. You can apply the function you want when computing the loss as @fengsam pointed out.

The only important thing to understand here is that whenever writing some custom loss/accuracy function, `y_pred` will be logits.

**Comment by bonlime** [comment]
Hi, 
You're right, there is no softmax in this model because the original model doesn't have it. If somebody will make a PR adding an option to add softmax, I'll accept it

**Comment by bonlime** [comment]
@YanZhiyuan0918  in this model you need label to have shape `(input_size, input_size, number_of_classes) `. If you want to use softmax, you need a background class, if you want to use a sigmoid activation on logits, you don't need a background

**Comment by JeniaNovellusDx** [comment]
Just add a custom layer in the end:

x = Activation('softmax')(x)

model = Model(input, x)

**Comment by YanZhiyuan0918** [comment]
@bonlime  Thanks for your reply! But I'm still confused about the label processing. According to your reply, for pascal voc dataset, the label shape should be (513, 513, 21), right? Could you please give some example codes to translate the original voc label into this shape? Thanks!

**Comment by bonlime** [comment]
@YanZhiyuan0918 you can look in original TF Deeplab repo, they have all the code you need

**Comment by bonlime** [comment]
added option to add activation while initiating the model

**Comment by SorourMo** [comment]
> @bonlime Thanks for your reply! But I'm still confused about the label processing. According to your reply, for pascal voc dataset, the label shape should be (513, 513, 21), right? Could you please give some example codes to translate the original voc label into this shape? Thanks!

If you use "sparse_categorical_crossentropy" as your loss function you do not need that conversion (to 21 channels).
